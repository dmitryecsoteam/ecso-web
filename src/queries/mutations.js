import gql from 'graphql-tag';

export const SIGNUP_USER = gql`
mutation signupUser($email: String, $password: String, $name: String){
    signupUser (email: $email, password: $password, name: $name) {
        token
    }
}
`;