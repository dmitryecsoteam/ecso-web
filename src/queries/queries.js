import gql from 'graphql-tag';

export const ORIGIN_INPUT_SEARCH = gql`
query originStartsWith($name: String){
    originStartsWith (name: $name) {
        name,
        country
    }
}
`