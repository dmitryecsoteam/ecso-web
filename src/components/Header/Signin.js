import React from 'react';
import { Mutation } from 'react-apollo';

import InputWithErrorTooltip from '../MainPage/InputWithErrorTooltip';
import { SIGNIN_USER } from '../../queries/mutations';


export default class Signin extends React.Component {
    state = {
        email: '',
        password: '',
        emailError: false,
        emailErrorText: '',
        passwordError: false,
        passwordErrorText: ''
    }

    onInputChange = (event) => {
        const { name, value } = event.target;

        this.setState({
            [name]: value
        });
    }

    onFormSubmit = async (event, signinUser) => {
        event.preventDefault();

        let emailError = false,
            emailErrorText = '',
            passwordError = false,
            passwordErrorText = '';

            const { email, password } = this.state;

        if (!email) {
            emailError = true;
            emailErrorText = 'Enter your email';
        }

        if (!password) {
            passwordError = true;
            passwordErrorText = 'Enter password';
        }

        this.setState({
            emailError,
            emailErrorText,
            passwordError,
            passwordErrorText
        });

        if (!(emailError || passwordError)) {
            try {

                const data = await signinUser();
                console.log(data)

            } catch (e) {

                if (e.message.includes('User with email')) {
                    this.setState({
                        emailError: true,
                        emailErrorText: e.message.replace('GraphQL error: ', '')
                    });
                }

                if (e.message.includes('Wrong password')) {
                    this.setState({
                        passwordError: true,
                        passwordErrorText: e.message.replace('GraphQL error: ', '')
                    });
                }
            }
        }
    }

    render() {

        const {
            email,
            password,
            emailError,
            emailErrorText,
            passwordError,
            passwordErrorText
        } = this.state;

        return (
            <div>
                <h2 className="signup__title">Signin with email</h2>

                <Mutation mutation={SIGNIN_USER} variables={{ email, password }}>

                    {(signinUser, { data, loading, error }) => {

                        return (
                            <form
                                onSubmit={e => this.onFormSubmit(e, signinUser)}
                                autoComplete="off"
                            >

                                <InputWithErrorTooltip
                                    error={emailError}
                                    errorText={emailErrorText}
                                >
                                    <input
                                        type="text"
                                        name="email"
                                        value={email}
                                        placeholder="Email"
                                        className="signup__input"
                                        onChange={this.onInputChange}
                                    />
                                </InputWithErrorTooltip>

                                <InputWithErrorTooltip
                                    error={passwordError}
                                    errorText={passwordErrorText}
                                >
                                    <input
                                        type="password"
                                        name="password"
                                        value={password}
                                        placeholder="Password"
                                        className="signup__input"
                                        onChange={this.onInputChange}
                                    />
                                </InputWithErrorTooltip>

                                <button className="search-form__submit-btn">Signin</button>
                            </form>
                        )
                    }}

                </Mutation>
            </div>
        );
    }
}