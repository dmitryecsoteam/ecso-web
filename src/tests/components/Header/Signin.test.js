import React from 'react';
import { mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import { GraphQLError } from 'graphql';

import Signin from '../../../components/Header/Signin';
import { SIGNIN_USER } from '../../../queries/mutations';



const waitForAsync = () => new Promise(resolve => setTimeout(resolve, 0));

let wrapper;

beforeEach(() => {
    const mocks = [{
        request: {
            query: SIGNIN_USER,
            variables: { email: 'test@mail.com', password: 'asdas' }
        },
        result: { data: { signinUser: { token: 'qwerty12' } } }
    }, {
        request: {
            query: SIGNIN_USER,
            variables: { email: 'doesntExist@mail.com', password: 'asdas' }
        },
        result: {
            errors: [new GraphQLError('User with email doesntExist@mail.com doesn\'t exist')],
        }
    }, {
        request: {
            query: SIGNIN_USER,
            variables: { email: 'test@mail.com', password: 'wrongPassword' }
        },
        result: {
            errors: [new GraphQLError('Wrong password')],
        }
    }];

    wrapper = mount(
        <MockedProvider mocks={mocks} addTypename={false}>
            <Signin />
        </MockedProvider>
    );
});

test('should render Signin component', () => {
    expect(wrapper.find(Signin)).toMatchSnapshot();
});

test('should signin user and return token', async () => {

    wrapper.find('[name="email"]').simulate('change', { target: { name: 'email', value: 'test@mail.com' } });
    wrapper.find('[name="password"]').simulate('change', { target: { name: 'password', value: 'asdas' } });

    wrapper.find('form').simulate('submit');
    await waitForAsync();
});

test('should not signin user if email doesn\'t exist', async () => {

    wrapper.find('[name="email"]').simulate('change', { target: { name: 'email', value: 'doesntExist@mail.com' } });
    wrapper.find('[name="password"]').simulate('change', { target: { name: 'password', value: 'asdas' } });

    wrapper.find('form').simulate('submit');
    await waitForAsync();

    const signin = wrapper.find(Signin);
    // Wait for setState to update
    await waitForAsync();

    expect(signin.state('emailError')).toBeTruthy;
    expect(signin.state('emailErrorText')).toEqual('User with email doesntExist@mail.com doesn\'t exist');
});

test('should not signin user if password is wrong', async () => {

    wrapper.find('[name="email"]').simulate('change', { target: { name: 'email', value: 'test@mail.com' } });
    wrapper.find('[name="password"]').simulate('change', { target: { name: 'password', value: 'wrongPassword' } });

    wrapper.find('form').simulate('submit');
    await waitForAsync();

    const signin = wrapper.find(Signin);
    // Wait for setState to update
    await waitForAsync();

    expect(signin.state('passwordError')).toBeTruthy;
    expect(signin.state('passwordErrorText')).toEqual('Wrong password');
});