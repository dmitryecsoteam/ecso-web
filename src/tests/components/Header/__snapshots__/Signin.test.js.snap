// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Signin component 1`] = `
<Signin>
  <div>
    <h2
      className="signup__title"
    >
      Signin with email
    </h2>
    <Mutation
      mutation={
        Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": Object {
                "kind": "Name",
                "value": "signinUser",
              },
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "email",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "email",
                          },
                        },
                      },
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "password",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "password",
                          },
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "signinUser",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "token",
                          },
                          "selectionSet": undefined,
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [
                Object {
                  "defaultValue": undefined,
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "String",
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "email",
                    },
                  },
                },
                Object {
                  "defaultValue": undefined,
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "String",
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "password",
                    },
                  },
                },
              ],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 133,
            "start": 0,
          },
        }
      }
      variables={
        Object {
          "email": "",
          "password": "",
        }
      }
    >
      <form
        autoComplete="off"
        onSubmit={[Function]}
      >
        <_default
          error={false}
          errorText=""
        >
          <div
            className="search-form__input-container"
          >
            <div
              className="search-form__hint"
            />
            <div
              className="search-form__label-input"
            >
              <input
                className="signup__input"
                name="email"
                onChange={[Function]}
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
            <div
              className="search-form__error"
            >
              <span
                className="search-form__error-text"
              />
            </div>
          </div>
        </_default>
        <_default
          error={false}
          errorText=""
        >
          <div
            className="search-form__input-container"
          >
            <div
              className="search-form__hint"
            />
            <div
              className="search-form__label-input"
            >
              <input
                className="signup__input"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              className="search-form__error"
            >
              <span
                className="search-form__error-text"
              />
            </div>
          </div>
        </_default>
        <button
          className="search-form__submit-btn"
        >
          Signin
        </button>
      </form>
    </Mutation>
  </div>
</Signin>
`;
